tree/ntuple/v7/src/RField.cxx

```diff
@@ -3218,7 +3235,18 @@ ROOT::Experimental::RVariantField::RVariantField(std::string_view fieldName,
       fTraits &= itemFields[i]->GetTraits();
       Attach(std::unique_ptr<RFieldBase>(itemFields[i]));
    }
-   fTagOffset = fMaxItemSize;
+
+   // With certain template parameters, the union of members of an std::variant starts at an offset > 0.
+   // For instance, std::variant<std::optional<int>> on macOS.
+   auto cl = TClass::GetClass(GetTypeName().c_str());
+   assert(cl);
+   auto dm = reinterpret_cast<TDataMember *>(cl->GetListOfDataMembers()->First());
+   if (dm)
+      fVariantOffset = dm->GetOffset();
+
+   const auto tagSize = GetVariantTagSize();
+   const auto padding = tagSize - (fMaxItemSize % tagSize);
```

Comment:
If `tagSize` is always a power of two the modulo operation can be written more efficiently as `fMaxItemSize & (tagSize - 1)`

Reply:
I would advise against this kind of micro-optimizations: Field creation is not on the critical path and we have lots of `virtual` function calls and interpreter invocations here. I also doubt that (integer) modulo operations are still significantly slower than subtraction + bit operation on modern architectures, except for microbenchmarks (maybe). IMHO readability and maintainability should be prioritized.

Reply:
I tend to agree here, I'd prioritize maintainability.